R/autosyn.R

```diff
@@ -472,3 +472,200 @@ flywire_ntplot3d <- function(x, nts=c("gaba", "acetylcholine", "glutamate",
+                                                ...){
+  method = match.arg(method)
+  rootid = x$flywire.id
+  if(is.null(connectors)){
+    synapses = flywire_partners(rootid = rootid,
+                                partners = "both",
+                                roots=TRUE,
+                                details=TRUE,
+                                cloudvolume.url = cloudvolume.url,
+                                method=method,
+                                Verbose=Verbose,
+                                local = local,
+                                ...)
+    if(isTRUE(!is.null(synapses) && !nrow(synapses))){
+      class(x) = union(c("flywireneuron", "catmaidneuron"), class(x))
+      return(x)
+    }
+    if(remove_autapses) {
+      synapses=synapses[synapses$post_id!=synapses$pre_id,,drop=FALSE]
+    }
+    # Add synapses
+    synapses %>% dplyr::group_by(.data$prepost) %>%
+      dplyr::filter(.data$cleft_scores>cleft.threshold) %>%
+      dplyr::mutate(x = ifelse(rootid==.data$pre_svid, .data$pre_x, .data$post_x)) %>%
+      dplyr::mutate(y = ifelse(rootid==.data$pre_svid, .data$pre_y, .data$post_y)) %>%
+      dplyr:: mutate(z = ifelse(rootid==.data$pre_svid, .data$pre_z, .data$post_z)) %>%
+      dplyr:: arrange(desc(.data$scores),desc(.data$cleft_scores)) %>%
+      dplyr::select(.data$offset, .data$prepost, .data$x, .data$y, .data$z,
+               .data$scores, .data$cleft_scores,
+               .data$segmentid_pre, .data$segmentid_post, .data$pre_svid, .data$post_svid,
+               .data$pre_id, .data$post_id) %>%
+      as.data.frame() ->
+      synapses.xyz
+  }else{
+    synapses=connectors[connectors$post_id%in%rootid|connectors$pre_id%in%rootid,,drop=FALSE]
+  }
+  # If transmiters
+  if(transmitters){
+    if(Verbose){
+      message("Adding transmitter prediction information (Eckstein et al. 2020)")
+    }
+    npred = flywire_ntpred(x=synapses.xyz, local = local, cloudvolume.url = cloudvolume.url)
+    pref.order = c("offset", "x", "y", "z", "scores", "cleft_scores", "top.p", "top.nt", "gaba", "acetylcholine",
+                   "glutamate", "octopamine", "serotonin", "dopamine", "prepost",
+                   "segmentid_pre", "segmentid_post",
+                   "pre_svid", "post_svid", "pre_id", "post_id")
+    pref.order = intersect(pref.order,colnames(npred))
+    if(nrow(npred)){
+      synapses.xyz = npred[,pref.order]
+    }
+  }else{
+    synapses.xyz$top.nt = "unknown"
+  }
+  near = nabor::knn(query= nat::xyzmatrix(synapses.xyz),data=nat::xyzmatrix(x$d),k=1)
+  synapses.xyz$treenode_id = x$d[near$nn.idx,"PointNo"]
+  # synapses.xyz = synapses.xyz[near$nn.dists<10000,] # remove erroneously associated synapses
+  synapses.xyz$connector_id = synapses.xyz$segmentid_pre
+  x$connectors = as.data.frame(synapses.xyz, stringsAsFactors = FALSE)
+  x$transmitter.predictions = table(subset(synapses.xyz, synapses.xyz$prepost == 0)$top.nt)
```

Comment:
@alexanderbates I think this is potentially misleading. It is the prediction for all synapses irrespective of whether they are input or output and will therefore not be the same as running flywire_ntpred on the neuron, which only considers pre-synapses for prediction. I guess you cold have separate summaries for outputs and inputs.

Reply:
`synapses.xyz$prepost == 0` should mean we are just considering output synapses for the neuron?

Reply:
OK. Sorry. That's odd because it doesn't give the same results as e.g. `flywire_ntpred('720575940634367221')`. Can you check?

Reply:
So one reason is that `flywire_ntpred` does not remove autapses. Should its default behaviour be to do this?

Reply:
And should this be part of this PR?

Reply:
The other reason is filtering `cleft_score`. Perhaps we should enable `flywire_partners` to do this (and optionally remove autapses when `roots=TRUE`), and carry the arguments through when using the `flywire_ntpred` function.

Reply:
Can you try to use `flywire_ntpred` directly here rather than have a separate logic? You can feed it the synapse dataframe, but you might need to give it the extra logic to inspect the prepost column if it exists and use it to select only the presynapses. It would make sense to remove autapses, I agree, so please go ahead and add the option to `flywire_ntpred`. I had seen `flywire_partners` very much as basic function that just returns "raw" data for others to use so I would try to avoid adding too much special logic there if possible.

Reply:
So right now, `flywire_neurons_add_synapses.neuron` calls `flywire_ntpred`, giving it `synapses.xyz`. Afterwards, I take only `prepost==0` to give a majority prediction. I can add a `cleft_score` threshold and `autapses` argument to  `flywire_ntpred` in this PR.